### ISSUE TITLE
Java UDP socket creation fails in LX zone (CentOS 6.6 lx-brand 20150415)

### ISSUE THREAD

**Issue by pannon** [joyent/smartos-live#446]
Tested with ancient Jboss4 and java 1.5 (third party black-box app). Jboss/Java is unable to create a UDP sockets.

LX image is: CentOS 6.6 lx-brand 20150415
Platform image: 20150320T005835Z

Part of the logs:

```
Caused by: java.lang.Exception: problem creating sockets (bind_addr=hostnamexxx/192.168.105.238, mcast_addr=null)
        at org.jgroups.protocols.UDP.start(UDP.java:372)
        at org.jgroups.stack.Protocol.handleSpecialDownEvent(Protocol.java:571)
        ... 135 more
Caused by: java.net.SocketException: Invalid argument
        at java.net.PlainDatagramSocketImpl.setTimeToLive(Native Method)
        at java.net.MulticastSocket.setTimeToLive(MulticastSocket.java:209)
        at org.jgroups.protocols.UDP.createSockets(UDP.java:459)
        at org.jgroups.protocols.UDP.start(UDP.java:367)
```

Happy to test/diagnose this further.

**Comment by pfmooney** [comment]
If it's not too much effort, an strace of the troublesome code would be useful.
I suspect it's attempting to do something like perform a setsockopt we don't yet support.

**Comment by pannon** [comment]
strace returned the following message: `ERROR: tracer already exists`

Used the following dtrace script instead:

```
#!/native/usr/sbin/dtrace -s
 
#pragma D option quiet
 
/**
PROTO values can be found here:
https://en.wikipedia.org/wiki/List_of_IP_protocol_numbers
SOCKOPT values can be found here:
https://github.com/joyent/illumos-joyent/blob/master/usr/src/lib/brand/lx/lx_brand/common/socket.c
*/
BEGIN
{
        printf("%6s %20s %5s %s\n", "PID", "NAME", "PROTO", "SOCKOPT");
}
 
lx-syscall::setsockopt:entry
{
        printf("%6d %20s %5d %d\n", pid, execname, arg1, arg2);
}
```

Output:

```
   PID                 NAME PROTO SOCKOPT
 79420                 java     0 1
 79420                 java     0 33
 79420                 java     1 2
 79420                 java     1 6
 79420                 java     1 7
 79420                 java     1 8
 79420                 java     1 9
 79420                 java     6 1
```

**Comment by pfmooney** [comment]
Can you gather the return codes too with `lx-syscall::setsockopt:return` and `arg1`?

**Comment by pannon** [comment]
```
   PID                 NAME PROTO SOCKOPT RETURN
 84711                 java     1 6         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 6         0
 84711                 java     1 6         0
 84711                 java     0 1         0
 84711                 java     1 6         0
 84711                 java     1 2         0
 84711                 java     0 33       -22
 84711                 java     1 6         0
 84711                 java     1 8         0
 84711                 java     1 7         0
 84711                 java     1 6         0
 84711                 java     1 8         0
 84711                 java     1 7         0
 84711                 java     1 6         0
 84711                 java     1 8         0
 84711                 java     1 7         0
 84711                 java     1 6         0
 84711                 java     1 6         0
 84711                 java     0 1         0
 84711                 java     1 6         0
 84711                 java     1 2         0
 84711                 java     0 33       -22
 84711                 java     1 2         0
 84711                 java     1 2         0
 84711                 java     1 6         0
 84711                 java     1 6         0
 84711                 java     0 1         0
 84711                 java     1 6         0
 84711                 java     1 2         0
 84711                 java     0 33       -22
 84711                 java     1 6         0
 84711                 java     1 6         0
 84711                 java     0 1         0
 84711                 java     1 6         0
 84711                 java     1 2         0
 84711                 java     0 33       -22
 84711                 java     1 2         0
 84711                 java     6 1         0
 84711                 java     1 9         0
 84711                 java     6 1         0
 84711                 java     1 2         0
```

**Comment by pfmooney** [comment]
It's having issues setting IP_MULTICAST_TTL which is supported by the java stacktrace.
I'll try to whip up a test case tomorrow.

**Comment by pannon** [comment]
Thanks @pfmooney , let me know if you need more testing from my end.

**Comment by pfmooney** [comment]
Could you tweak your dtrace script to also log arg4 (as %d) for setsockopt:entry?
Linux is more tolerant of length when it comes to sockopts and I have a hunch that java might be using an 'int' where a 'char' is expected.

**Comment by bahamat** [comment]
@pannon The message from `strace` is because you ran the native SmartOS `strace`, which is entirely different from what Linux calls `strace`.

**Comment by pannon** [comment]
Yeah, thanks @bahamat - I have figured that one.. but dtrace was already working by that time.

PS: don't you love the Linux double-speak.. ;)

**Comment by pannon** [comment]
@pfmooney please see below with modified ARG4:

```
   PID                 NAME PROTO SOCKOPT ARG4 RETURN
 50270                 java     1     6   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     6   4         0
 50270                 java     1     6   4         0
 50270                 java     0     1   4         0
 50270                 java     1     6   4         0
 50270                 java     1     2   4         0
 50270                 java     0    33   1       -22
 50270                 java     1     6   4         0
 50270                 java     1     8   4         0
 50270                 java     1     7   4         0
 50270                 java     1     6   4         0
 50270                 java     1     8   4         0
 50270                 java     1     7   4         0
 50270                 java     1     6   4         0
 50270                 java     1     8   4         0
 50270                 java     1     7   4         0
 50270                 java     1     6   4         0
 50270                 java     1     6   4         0
 50270                 java     0     1   4         0
 50270                 java     1     6   4         0
 50270                 java     1     2   4         0
 50270                 java     0    33   1       -22
 50270                 java     1     2   4         0
 50270                 java     1     2   4         0
 50270                 java     1     6   4         0
 50270                 java     1     6   4         0
 50270                 java     0     1   4         0
 50270                 java     1     6   4         0
 50270                 java     1     2   4         0
 50270                 java     0    33   1       -22
 50270                 java     1     6   4         0
 50270                 java     1     6   4         0
 50270                 java     0     1   4         0
 50270                 java     1     6   4         0
 50270                 java     1     2   4         0
 50270                 java     0    33   1       -22
 50270                 java     1     2   4         0
 50270                 java     6     1   4         0
 50270                 java     1     9   4         0
 50270                 java     6     1   4         0
 50270                 java     1     2   4         0
```

**Comment by pfmooney** [comment]
Open as [OS-4306](https://smartos.org/bugview/OS-4306).

**Comment by pfmooney** [comment]
This should be fixed in joyent/illumos-joyent@70fe4e726e40ffd9c0eabfcc9a4477c54d181a63
